### PR TITLE
[PATCH API-NEXT v20] IPsec implementation based on packet interface

### PR DESCRIPTION
**Pull request by lumag**
This PR replaces #28 by providing new IPsec implementation based on latest API additions.

### CODE REVIEW

**Comment by Bill-Fischofer-Linaro** [comment]
@lumag please rebase this as part of your next iteration since we've gotten into trouble trying to do complex rebasing merges in the past. Thanks.

**Comment by lumag** [comment]
@Bill-Fischofer-Linaro @muvarov I will rebase after #70 gets merged.

**Comment by muvarov** [comment]
@lumag some checkpatch errors still can be cleaned up.

**Comment by Bill-Fischofer-Linaro** [comment]
The only checkpatch items that appear to be legit are:

```
CHECK: Please don't use multiple blank lines
#124: FILE: platform/linux-generic/odp_packet.c:2456:
+
+
total: 0 errors, 0 warnings, 1 checks, 124 lines checked
NOTE: Ignored message types: AVOID_EXTERNS BIT_MACRO COMPARISON_TO_NULL DEPRECATED_VARIABLE NEW_TYPEDEFS PREFER_PRINTF PREFER_SCANF SPLIT_STRING SSCANF_TO_KSTRTO VOLATILE
0005-linux-gen-packet-support-parsing-of-just-L3-L4-heade.patch has style problems, please review.
```
and

```
HECK: Blank lines aren't necessary before a close brace '}'
#584: FILE: platform/linux-generic/odp_ipsec.c:511:
+
+	}
total: 0 errors, 0 warnings, 1 checks, 1323 lines checked
NOTE: Ignored message types: AVOID_EXTERNS BIT_MACRO COMPARISON_TO_NULL DEPRECATED_VARIABLE NEW_TYPEDEFS PREFER_PRINTF PREFER_SCANF SPLIT_STRING SSCANF_TO_KSTRTO VOLATILE
0009-linux-generic-ipsec-draft-IPsec-implementation.patch has style problems, please review.
```
The rest seem to be false-positives or checkpatch being overly pedantic.

**Comment by lumag** [comment]
fixed

File: platform/linux-generic/odp_ipsec.c [hunk 1]
```diff
@@ -4,179 +4,1244 @@
...
+	ip_frag_offset = 0;
+
+	ipsec_sa = _odp_ipsec_sa_use(sa);
+	ODP_ASSERT(NULL != ipsec_sa);
+
+	/* Initialize parameters block */
+	memset(&param, 0, sizeof(param));
+
+	if (ODP_IPSEC_MODE_TRANSPORT == ipsec_sa->mode &&
+	    _ODP_IPV4HDR_IS_FRAGMENT(ip->frag_offset)) {
```
Comment (JannePeltonen) [inline platform/linux-generic/odp_ipsec.c]: _ODP_IPV4HDR_IS_FRAGMENT() expects the parameter word in host byte order. The conversion is missing here.

File: platform/linux-generic/odp_ipsec.c [hunk 2]
```diff
@@ -4,179 +4,1244 @@
...
+	ip_tos = 0;
+	ip_ttl = 0;
+	ip_frag_offset = 0;
+
+	/* Initialize parameters block */
+	memset(&param, 0, sizeof(param));
+
+	ipsec_offset = ip_offset + ip_hdr_len;
+
+	if (_ODP_IPV4HDR_IS_FRAGMENT(ip->frag_offset)) {
```
Comment (JannePeltonen) [inline platform/linux-generic/odp_ipsec.c]: _ODP_IPV4HDR_IS_FRAGMENT() expects the parameter word in host byte order. The conversion is missing here.

Comment (lumag) [reply platform/linux-generic/odp_ipsec.c]: @JannePeltonen thank you, fixed both entries.

**Comment by muvarov** [comment]
These checkpatch warnings also can be fixed. Some of warnings very strange. Doxygen  error should go away after rebase.

WARNING: line over 80 characters

#622: FILE: platform/linux-generic/odp_ipsec.c:548:

+		packet_parse_l3_l4(pkt_hdr, parse_layer(ipsec_config.inbound.parse),

CHECK: Alignment should match open parenthesis

#2565: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:18:

+KEY(key_a5_128, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5,

+		0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5);

CHECK: Alignment should match open parenthesis

#2567: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:20:

+KEY(key_5a_128, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a,

+		0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a);

CHECK: Alignment should match open parenthesis

#2569: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:22:

+KEY(key_a5_256, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5,

+		0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5,

CHECK: Alignment should match open parenthesis

#2573: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:26:

+KEY(key_5a_256, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a,

+		0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a, 0x5a,

CHECK: Alignment should match open parenthesis

#2578: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:31:

+KEY(key_rfc3602, 0x90, 0xd3, 0x82, 0xb4, 0x10, 0xee, 0xba, 0x7a,

+		 0xd9, 0x38, 0xc4, 0x6c, 0xec, 0x1a, 0x82, 0xbf);

CHECK: Alignment should match open parenthesis

#2580: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:33:

+KEY(key_rfc3602_2, 0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,

+		   0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef);

CHECK: Alignment should match open parenthesis

#2582: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:35:

+KEY(key_mcgrew_gcm_2, 0xfe, 0xff, 0xe9, 0x92, 0x86, 0x65, 0x73, 0x1c,

+		      0x6d, 0x6a, 0x8f, 0x94, 0x67, 0x30, 0x83, 0x08);

CHECK: Alignment should match open parenthesis

#2585: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:38:

+KEY(key_mcgrew_gcm_3, 0xab, 0xbc, 0xcd, 0xde, 0xf0, 0x01, 0x12, 0x23,

+		      0x34, 0x45, 0x56, 0x67, 0x78, 0x89, 0x9a, 0xab,

CHECK: Alignment should match open parenthesis

#2590: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:43:

+KEY(key_mcgrew_gcm_4, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,

+		      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00);

CHECK: Alignment should match open parenthesis

#2593: FILE: test/common_plat/validation/api/ipsec/test_vectors.h:46:

+KEY(key_mcgrew_gcm_12, 0x7d, 0x77, 0x3d, 0x00, 0xc1, 0x44, 0xc5, 0x25,

+		       0xac, 0x61, 0x9d, 0x18, 0xc8, 0x4a, 0x3f, 0x47);

**Comment by lumag** [comment]
@muvarov no, there is no way to fix these `Alignment should match open parenthesis` checks. Will fix 80-chars limit though. BTW: does anybody still use 80x25 terminals? There is no other reason to impose 80-char line length limit.

**Comment by muvarov** [comment]
@lumag looks like some bug in checkpatch. I though that change like this will help:
https://github.com/muvarov/odp/commit/f9103cc36e882e6401bd907c6af0a1e3325a6899

but it generates:
WARNING: please, no spaces at the start of a line

**Comment by lumag** [comment]
@muvarov it is just checkpatch not expecting that the indentation consists only of several spaces.

**Comment by Bill-Fischofer-Linaro** [comment]
Doxygen error should be fixed by a rebase. Looks like it's still running without Maxim's latest fix. Checkpatch errors seem to be false positives.

**Comment by muvarov** [comment]
@lumag merging as is or you will update it?

**Comment by lumag** [comment]
@muvarov I will update this PR anyway, have some small fixes here and there.

**Comment by lumag** [comment]
@muvarov PR updated up to the latest version

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**